File: hahlw/array/74.搜索二维矩阵.cpp
/*
 * @lc app=leetcode.cn id=74 lang=cpp
 *
 * [74] 搜索二维矩阵
 */
class Solution {
public:
    //核心思想两次二分，相比从左下角开始遍历时间复杂度要少。
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.empty()||matrix[0].empty())return false;
        int m = matrix.size();
        int n =matrix[0].size();

        int lo=0,hi=m-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(matrix[mid][0]==target)return true;
            else if(matrix[mid][0]>target)hi=mid-1;
            else lo=mid+1;
        }
        if(hi<0)return false;//如果小于零，那么说明target比最小值还要小，即不存在。
        int row=hi;
        lo=0,hi=n-1;
        while(lo<=hi){
            int mid=lo+(hi-lo)/2;
            if(matrix[row][mid]==target)return true;
            else if(matrix[row][mid]>target)hi=mid-1;
            else lo=mid+1;
        }
        return false;
    }
};

File: Juniper/74.搜索二维矩阵.py
#
# @lc app=leetcode.cn id=74 lang=python3
#
# [74] 搜索二维矩阵
#
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        if not matrix:
            return False
        row = len(matrix)
        col = len(matrix[0])     
        # # 5.79 %
        left = 0
        right = col*row-1
        while left <= right :
            mid = (left + right)//2
            index1 = mid // col 
            index2 =  mid % col
            if target < matrix[index1][index2]:
                right = mid -1
            elif target > matrix[index1][index2]:
                left = mid +1
            else :
                return True
        return False
        # # 5.79 %
        # i = 0
        # j = col-1
        # while i < row and j >= 0:
        #     if matrix[i][j] < target:
        #         i +=1
        #     elif matrix[i][j] > target:
        #         j -=1
        #     else:
        #         return True
        # return False
    
       
  
   

